syntax = "proto3";

package replication;

service Sequence {
  rpc Write (WriteRequest) returns (WriteResponse);
  rpc Heartbeat (HeartbeatRequest) returns (HeartbeatResponse);
  rpc RequestVote (RequestVoteRequest) returns (RequestVoteResponse);
}

message WriteRequest {
  string key = 1;
  string value = 2;
}

message WriteResponse {
  string ack = 1;
}

message HeartbeatRequest {
  string service_identifier = 1; // Identifier for the service sending the heartbeat
}

message HeartbeatResponse {
  string message = 1; // Optional message in the response
}

message RequestVoteRequest {
  int32 term = 1;
  int32 candidate_id = 2;
  int32 last_log_index = 3;
  int32 last_log_term = 4;
}

message RequestVoteResponse {
  bool vote_granted = 1;
}